server/src/test/java/org/apache/kafka/network/SocketServerConfigsTest.java

```diff
@@ -0,0 +1,107 @@
+        expected.put(new ListenerName("PLAINTEXT"), SecurityProtocol.PLAINTEXT);
+        expected.put(new ListenerName("SSL"), SecurityProtocol.SSL);
+        expected.put(new ListenerName("SASL_PLAINTEXT"), SecurityProtocol.SASL_PLAINTEXT);
+        expected.put(new ListenerName("SASL_SSL"), SecurityProtocol.SASL_SSL);
+        assertEquals(expected, SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO);
+    }
+
+    @Test
+    public void testListenerListToEndPointsWithEmptyString() {
+        assertEquals(Arrays.asList(),
+            SocketServerConfigs.listenerListToEndPoints("",
+                SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    @Test
+    public void testListenerListToEndPointsWithBlankString() {
+        assertEquals(Arrays.asList(),
+            SocketServerConfigs.listenerListToEndPoints(" ",
+                SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    @Test
+    public void testListenerListToEndPointsWithOneEndpoint() {
+        assertEquals(Arrays.asList(
+            new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "example.com", 8080)),
+                SocketServerConfigs.listenerListToEndPoints("PLAINTEXT://example.com:8080",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    // Regression test for KAFKA-3719
+    @Test
+    public void testListenerListToEndPointsWithUnderscores() {
+        assertEquals(Arrays.asList(
+            new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "example.com", 8080),
+            new Endpoint("SSL", SecurityProtocol.SSL, "local_host", 8081)),
+                SocketServerConfigs.listenerListToEndPoints("PLAINTEXT://example.com:8080,SSL://local_host:8081",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    @Test
+    public void testListenerListToEndPointsWithWildcard() {
+        assertEquals(Arrays.asList(
+            new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, null, 8080)),
+                SocketServerConfigs.listenerListToEndPoints("PLAINTEXT://:8080",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    @Test
+    public void testListenerListToEndPointsWithIpV6() {
+        assertEquals(Arrays.asList(
+            new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "::1", 9092)),
+                SocketServerConfigs.listenerListToEndPoints("PLAINTEXT://[::1]:9092",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
+    }
+
+    @Test
+    public void testAnotherListenerListToEndPointsWithIpV6() {
+        assertEquals(Arrays.asList(
+            new Endpoint("SASL_SSL", SecurityProtocol.SASL_SSL, "fe80::b1da:69ca:57f7:63d8%3", 9092)),
+                SocketServerConfigs.listenerListToEndPoints("SASL_SSL://[fe80::b1da:69ca:57f7:63d8%3]:9092",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
```

Comment:
Interesting that the character `%` is allowed in IPv6: https://superuser.com/a/99753

Reply:
Yeah, it's a puzzling design decision, especially given that  `%` is not a URL-safe character.